homeassistant/components/ssdp/__init__.py

```diff
@@ -85,6 +88,59 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+_HaServiceInfoDescription = TypedDict(
+    "_HaServiceInfoDescription",
+    {
+        "x-homeassistant-matching-domains": set,
+    },
+    total=True,
+)
+
+
+class _SsdpDescriptionBase(TypedDict, total=True):
+    """Compulsory keys for SSDP info."""
+
+    ssdp_usn: str
+    ssdp_st: str
+
+
+class SsdpDescription(_SsdpDescriptionBase, total=False):
+    """SSDP info with optional keys."""
+
+    ssdp_location: str
+    ssdp_nt: str
+    ssdp_udn: str
+    ssdp_ext: str
+    ssdp_server: str
+
+
+class _UpnpDescriptionBase(TypedDict, total=True):
+    """Compulsory keys for UPnP info."""
+
+    UDN: str
+
+
+class UpnpDescription(_UpnpDescriptionBase, total=False):
+    """UPnP info with optional keys."""
+
+    deviceType: str
+    friendlyName: str
+    manufacturer: str
+    manufacturerURL: str
+    modelDescription: str
+    modelName: str
+    modelNumber: str
+    modelURL: str
+    serialNumber: str
+    serviceList: str
+    UPC: str
+    presentationURL: str
```

Comment:
Here's the keys and types derived from [UPnP Device Architecture v1.1](http://upnp.org/specs/arch/UPnP-arch-DeviceArchitecture-v1.1.pdf) section 2.3. The names match with what you have above, but the types and whether they're required are specified in a bit more detail.

Required (total=True):
```py
deviceType: str
friendlyName: str
manufacturer: str
modelName: str
UDN: str
```

Optional/recommended (total=False):
```py
manufacturerURL: str
modelDescription: str
modelNumber: str
modelURL: str
serialNumber: str
UPC: str
iconList: dict[str,list[dict[str,str]]]
serviceList: dict[str,list[dict[str,str]]]
deviceList: dict[str, Any]
presentationURL: str
```

This could go into further detail for the `iconList`, `serviceList`, and `deviceList`, but I didn't think it worth defining them with further TypedDicts.

Reply:
I have adjusted the descriptions to match your recommendations.